### PR TITLE
feat(calculation bcl convert metrics)

### PR DESCRIPTION
**Pull request by ChrOertlin**
## Description

The code represents a Python module for parsing BCL convert metrics data and converting it into sequencing statistics models. The module consists of three files: `bcl_convert.py`, `bcl_convert_to_sequencing_statistics.py`, and `bcl_convert_metrics.py`. Additionally, there are test fixtures defined in `conftest.py` for unit testing.

In the `bcl_convert.py` file, there is a class called `BclConvertMetricsParser`. This class is responsible for parsing BCL convert metrics data and providing methods to extract various sequencing statistics. The class takes a directory path containing BCL convert metrics files as input during initialization. It then reads and parses the quality metrics, demultiplex metrics, adapter metrics, sample sheet, and run info files.

The parsed metrics are stored as attributes in the class instance, including lists of quality metrics, demultiplex metrics, adapter metrics, sample sheet data, and run info. The class provides several methods to retrieve specific sequencing statistics based on sample internal IDs, lanes, and other parameters. Some of the provided methods include:

- `get_sample_internal_ids()`: Returns a list of sample internal IDs present in the demultiplex metrics.
- `get_lanes_for_sample(sample_internal_id)`: Returns a list of lane numbers associated with a given sample internal ID.
- `calculate_total_reads_for_sample_in_lane(sample_internal_id, lane)`: Calculates the total number of reads for a sample in a specific lane.
- `get_flow_cell_name()`: Returns the flow cell name mentioned in the sample sheet.
- `get_q30_bases_percent_for_sample_in_lane(sample_internal_id, lane)`: Calculates the percentage of bases that have a quality score of at least Q30 for a sample in a specific lane.
- `get_mean_quality_score_fot_sample_in_lane(sample_internal_id, lane)`: Returns the mean quality score (Q30) for a sample in a specific lane.

The `bcl_convert_to_sequencing_statistics.py` file defines a function `create_sample_lane_sequencing_metrics_from_bcl_convert_metrics` that utilizes the `BclConvertMetricsParser` class. This function takes the directory path of BCL convert metrics as input and uses the parser to extract sequencing statistics for each sample and lane. It returns a list of `SampleLaneSequencingMetrics` objects, which are models representing the sequencing statistics for a specific sample and lane.

The `bcl_convert_metrics.py` file contains constants related to the BCL convert metrics, such as file names and column names.

Lastly, the `conftest.py` file defines various fixtures for unit testing, including file paths and sample data.

Overall, this code provides a comprehensive solution for parsing BCL convert metrics and converting them into sequencing statistics models, allowing further analysis and processing of the sequencing data.

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py [hunk 1]
```diff
@@ -93,3 +93,75 @@ def parse_run_info_file(self) -> BclConvertRunInfo:
...
+        metric: BclConvertQualityMetrics = self.get_metrics_for_sample_internal_id_and_lane(
+            metrics_list=self.quality_metrics, sample_internal_id=sample_internal_id, lane=lane
+        )
+        return int(metric.yield_bases / 1000000)
+
+    def get_flow_cell_name(self) -> str:
+        """Return the flow cell name of the demultiplexed flow cell."""
+        return self.sample_sheet[0].flow_cell_name
+
+    def get_q30_bases_percent_per_lane(self, sample_internal_id, lane) -> float:
```
Comment (seallard) [inline cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py:146]: type hints for parameters

File: cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py [hunk 2]
```diff
@@ -93,3 +93,75 @@ def parse_run_info_file(self) -> BclConvertRunInfo:
...
+        """Return the percent of bases that are Q30 for a sample and lane."""
+        metric: BclConvertQualityMetrics = self.get_metrics_for_sample_internal_id_and_lane(
+            metrics_list=self.quality_metrics, sample_internal_id=sample_internal_id, lane=lane
+        )
+        return metric.q30_bases_percent
+
+    def get_perfect_index_reads_percent_for_sample_per_lane(
+        self, sample_internal_id, lane
```
Comment (seallard) [inline cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py:154]: type hints for parameters

File: cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py [hunk 3]
```diff
@@ -93,3 +93,75 @@ def parse_run_info_file(self) -> BclConvertRunInfo:
...
+    ) -> float:
+        """Return the percent of perfect index reads for a sample and lane."""
+        metric: BclConvertDemuxMetrics = self.get_metrics_for_sample_internal_id_and_lane(
+            metrics_list=self.demux_metrics, sample_internal_id=sample_internal_id, lane=lane
+        )
+        return metric.perfect_index_reads_percent
+
+    def get_mean_quality_score_per_lane(self, sample_internal_id, lane) -> float:
```
Comment (seallard) [inline cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py:162]: type hints for parameters

File: cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py [hunk 4]
```diff
@@ -93,3 +93,75 @@ def parse_run_info_file(self) -> BclConvertRunInfo:
...
+        self,
+        metrics_list: List[
+            Union[BclConvertQualityMetrics, BclConvertDemuxMetrics, BclConvertAdapterMetrics]
+        ],
+        sample_internal_id: str,
+        lane: int,
+    ) -> Union[BclConvertQualityMetrics, BclConvertDemuxMetrics, BclConvertAdapterMetrics]:
+        """Return the metrics for a sample by sample internal id."""
+        for metric in metrics_list:
+            if metric.sample_internal_id == sample_internal_id and metric.lane == lane:
```
Comment (seallard) [inline cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py:123]: There should only be one entry for a sample and lane? At least that is assumed here and I guess it makes sense!

Comment (ChrOertlin) [reply cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py:123]: Yes the other sample entry is the other lane

File: cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py [hunk 5]
```diff
@@ -93,3 +93,75 @@ def parse_run_info_file(self) -> BclConvertRunInfo:
         LOG.info(f"Parsing Run info XML {self.run_info_path}")
         parsed_metrics = BclConvertRunInfo(tree=ET.parse(self.run_info_path))
         return parsed_metrics
+
+    def get_sample_internal_ids(self) -> List[str]:
+        """Return a list of sample internal ids."""
+        sample_internal_ids: List[str] = []
+        for sample_demux_metric in self.demux_metrics:
+            if sample_demux_metric.sample_project not in ["indexcheck", "Undetermined"]:
```
Comment (seallard) [inline cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py:101]: Can this if check be extracted to a separate function with a descriptive name?

File: cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py [hunk 6]
```diff
@@ -93,3 +93,75 @@ def parse_run_info_file(self) -> BclConvertRunInfo:
...
+                sample_internal_ids.append(sample_demux_metric.sample_internal_id)
+        return sample_internal_ids
+
+    def get_lanes_for_sample_internal_id(self, sample_internal_id: str) -> List[int]:
```
Comment (seallard) [inline cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py:105]: Is it necessary to aggregate data for a sample across all lanes? I have not done that

File: cg/apps/sequencing_metrics_parser/parsers/bcl_convert_to_sequencing_statistics.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        ):
+            sequencing_statistics.append(
+                SequencingStatistics(
+                    sample_internal_id=sample_internal_id,
+                    flow_cell_name=metrics_parser.get_flow_cell_name(),
+                    lane=lane,
+                    yield_in_megabases=metrics_parser.calculate_total_yield_per_lane_in_mega_bases(
+                        sample_internal_id=sample_internal_id, lane=lane
+                    ),
+                    read_counts=metrics_parser.calculate_total_reads_per_lane(
```
Comment (seallard) [inline cg/apps/sequencing_metrics_parser/parsers/bcl_convert_to_sequencing_statistics.py:40]: `total_reads_for_sample_in_lane` or something similar might be a more descriptive name

File: cg/apps/sequencing_metrics_parser/parsers/bcl_convert_to_sequencing_statistics.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                    ),
+                    passed_filter_percent=DRAGEN_PASSED_FILTER_PCT,
+                    raw_clusters_per_lane_percent=0,
+                    perfect_index_reads_percent=metrics_parser.get_perfect_index_reads_percent_for_sample_per_lane(
+                        sample_internal_id=sample_internal_id, lane=lane
+                    ),
+                    bases_with_q30_percent=metrics_parser.get_q30_bases_percent_per_lane(
+                        sample_internal_id=sample_internal_id, lane=lane
+                    ),
+                    lanes_mean_quality_score=metrics_parser.get_mean_quality_score_per_lane(
```
Comment (seallard) [inline cg/apps/sequencing_metrics_parser/parsers/bcl_convert_to_sequencing_statistics.py:51]: We should rename this field to be in singular too

File: tests/apps/sequencing_metrics_parser/conftest.py [hunk 9]
```diff
@@ -72,6 +74,36 @@ def fixture_test_lane() -> int:
...
+
+
+@pytest.fixture(name="bcl_convert_test_mean_quality_score_per_lane")
+def fixture_bcl_convert_test_mean_quality_score() -> float:
+    return 36.15
+
+
+@pytest.fixture(name="bcl_convert_test_flow_cell_name")
+def fixture_bcl_convert_test_flow_cell_name() -> str:
+    return "HY7FFDRX2"
```
Comment (diitaz93) [inline tests/apps/sequencing_metrics_parser/conftest.py:104]: Is it possible that that there is already a fixture for this

**Review by diitaz93 — COMMENTED** [review]
Im missing the tests but looks good 👍

File: cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py (hunk 6 above)
Comment (ChrOertlin) [reply cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py:105]: It is not used for data aggregation just to get the number of lanes on which a sample has been run. This is then used to fill the sequencing statistics model.

File: tests/apps/sequencing_metrics_parser/conftest.py (hunk 9 above)
Comment (ChrOertlin) [reply tests/apps/sequencing_metrics_parser/conftest.py:104]: I doubt it, I took files randomly from hasta that I use as dummy

File: cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py:1]: Out of scope - will handle that in a separate PR for sample sheet reading

**Review by seallard — APPROVED** [review]

File: cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py [hunk 10]
```diff
@@ -93,3 +100,63 @@ def parse_run_info_file(self) -> BclConvertRunInfo:
...
+        return self.sample_sheet[0].flow_cell_name
+
+    def get_q30_bases_percent_per_lane(self, sample_internal_id: str, lane: int) -> float:
+        """Return the percent of bases that are Q30 for a sample and lane."""
+        metric: BclConvertQualityMetrics = self.get_metrics_for_sample_internal_id_and_lane(
+            metrics_list=self.quality_metrics, sample_internal_id=sample_internal_id, lane=lane
+        )
+        return metric.q30_bases_percent
+
+    def get_mean_quality_score_per_lane(self, sample_internal_id: str, lane: int) -> float:
```
Comment (seallard) [inline cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py:157]: `get_mean_quality_score_for_sample_in_lane` might be a better name

File: tests/apps/sequencing_metrics_parser/conftest.py [hunk 11]
```diff
@@ -72,6 +42,26 @@ def fixture_test_lane() -> int:
     return 1
 
 
+@pytest.fixture(name="bcl_convert_reads_for_test_sample")
+def fixture_bcl_convert_reads_for_test_sample() -> int:
+    return 15962796
+
+
+@pytest.fixture(name="bcl_convert_test_q30_bases_percent")
```
Comment (seallard) [inline tests/apps/sequencing_metrics_parser/conftest.py:50]: I don't like this, to hard code a calculated value and use it in a test to verify a result. Better to calculate the value in the fixture instead or calculate it here with values from the fixture.

Comment (ChrOertlin) [reply tests/apps/sequencing_metrics_parser/conftest.py:50]: It is not calculated in this parser - it comes directly from the output

**Pull request merged**